File: apps/cli/src/api/terms.ts
import type { PeopleSoftClient } from "./client.ts";
import { ENDPOINTS } from "./endpoints.ts";
import {
  type Term,
  extractPageState,
  buildTermSelectBody,
  parseTermsFromHtml,
  parseStrmFromHtml,
} from "./peoplesoft.ts";

export type { Term };

export async function listTerms(client: PeopleSoftClient): Promise<Term[]> {
  const res = await client.get(ENDPOINTS.termList);
  return parseTermsFromHtml(res.body as string);
}

export async function selectTerm(client: PeopleSoftClient, termIndex: number): Promise<string> {
  const res = await client.get(ENDPOINTS.termList);
  const state = extractPageState(res.body as string);

  const postRes = await client.post(ENDPOINTS.termList, {
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: buildTermSelectBody(state, termIndex),
    followRedirect: false,
    throwHttpErrors: false,
  });

  const strm = parseStrmFromHtml(postRes.body as string);
  if (!strm) throw new Error("Could not determine STRM from term selection response.");
  return strm;
}

File: apps/cli/src/api/terms.rs
use anyhow::Result;

use super::client::PeopleSoftClient;
use super::endpoints;
use super::peoplesoft::{
    build_term_select_body, extract_page_state, parse_strm_from_html, parse_terms_from_html, Term,
};

pub async fn list_terms(client: &PeopleSoftClient) -> Result<Vec<Term>> {
    let body = client.get(&endpoints::term_list()).await?;
    Ok(parse_terms_from_html(&body))
}

pub async fn select_term(client: &PeopleSoftClient, term_index: i64) -> Result<String> {
    let url = endpoints::term_list();
    let body = client.get(&url).await?;
    let state = extract_page_state(&body);
    let select_body = build_term_select_body(&state, term_index);
    let resp = client.post(&url, select_body).await?;
    let strm = parse_strm_from_html(&resp)
        .ok_or_else(|| anyhow::anyhow!("Could not parse STRM from response"))?;
    Ok(strm)
}
